Explain and test two pkg dependency routines

tldahlgren opened the pull request:
Added comments to explain that `install_dependency_symlinks` and `flatten_dependencies` are actually used.

Added a test that exercises the routines.

tldahlgren commented:
How can it claim the method is not covered when I added a test for it?  (Perhaps a namespace issue at play here?)

scheibelp commented on lib/spack/spack/package.py:
```diff
@@ -2373,6 +2377,8 @@ def use_cray_compiler_names():
     os.environ['F77'] = 'ftn'
 
 
+# Note this is a feature required for external packages maintained by at least
```
I think it might be useful to also explain the supported use case (if you happen to know).

tldahlgren replied:
FYI. I've asked Greg (@becker33) if there's anything I can add to the comment.

becker33 replied:
Some codes require that all of their third party libraries are installed in a single location. Calling this in the install method can create that location.

tldahlgren replied:
I've iterated on the comment based on feedback.

scheibelp commented on lib/spack/spack/test/install.py:
```diff
@@ -139,6 +141,24 @@ def test_installed_dependency_request_conflicts(
+    assert dependency_name not in os.listdir(pkg.prefix)
+
+    install_dependency_symlinks(pkg, dependent, pkg.prefix)
+    assert dependency_name in os.listdir(pkg.prefix)
```
(minor) I think you can remove this check because the next assert is more specific and I think provides nearly as much info as the two of them together.

tldahlgren replied:
Fair enough.

tldahlgren replied:
Removed.

tldahlgren commented on lib/spack/spack/package.py:
```diff
@@ -2360,6 +2360,10 @@ class Package(PackageBase):
+# Note this is a feature required for external packages maintained by at least
+# one site.
+#
+# TODO: Why is unused pkg one of the arguments?
```
Okay.  I've asked Greg (@becker33) to confirm it's safe to make this change.

becker33 replied:
`pkg` is in the place of `self`, I must have had the intention of making it a static or class method at some point in the past.

tgamblin replied:
It's not an instance method where `self` would make sense -- we should remove `pkg`.

tldahlgren replied:
Removing `pkg` is not appropriate given the purpose of the method so I added a package to illustrate its use and associated documentation.

tldahlgren commented:
@scheibelp I added  comments based on @becker33's input (wrt use cases) and removed the assert per your feedback so this PR is ready for re-review.

tgamblin commented on lib/spack/spack/package.py:
```diff
@@ -2373,6 +2377,8 @@ def use_cray_compiler_names():
     os.environ['F77'] = 'ftn'
 
 
+# Note this is a feature required for external packages maintained by at least
```
It's going to clutter the code to have a bunch of meta-conversation (this is used by n sites, etc.) in the comments.

I think notes like this should go in the tests for the feature.  Or, we should just add an explicit test to check that features like this (things needed by external projects) exist.  e.g. `test_externally_required_calls_are_present`.  This would keep the code cleaner.

tldahlgren replied:
I added the comment to avoid the problem where a contributor removes a feature (without bothering to ask) that was not being explicitly used in the code base.  (Apparently it's happened on an actual code.)

tgamblin replied:
Yes, I'm just saying a better way to enforce that is to add a test.

tldahlgren replied:
How about changing the current test to use a (new) built-in mock that uses the install function as Greg explained?  Or is that what you are suggesting?

tldahlgren commented:
@tgamblin I (roughly) restored a comment I thought I'd added; removed those you didn't want; and changed the test to conform to the expected use of the function.  Is this version more acceptable?

tldahlgren commented:
@scheibelp I removed the assert you requested.

@tgamblin, @scheibelp, @becker33 This PR is ready to be reviewed again.

tldahlgren commented:
@tgamblin @becker33 @scheibelp Are you okay with how I addressed your feedback and or comment changes?

tldahlgren commented:
@tgamblin Finished the tweaks we discussed w.r.t. restoring `pkg` as the argument and removing `spack.package.` from the `install` assignment in the `no-source/package.py`.  Unfortunately, it appears the latter change negatively affect code coverage.

tldahlgren commented:
Interesting.  Looks like code coverage didn't recognize the package install.  So added an explicit `flatten_dependencies` test and performed (hopefully) remaining cleanup.

@scheibelp @tgamblin This PR is ready for another review round.
